src/com/ecfeed/core/generators/algorithms/NwiseScoreEvaluator_Rep.java

```diff
@@ -0,0 +1,151 @@
+
+	}
+
+	public int getdimensionCount() {
+		return dimensionCount;
+	}
+
+	private boolean plus(int[] encode, int[] range) {
+		for (int j = encode.length - 1; j >= 0; j--) {
+			if (encode[j] + 1 == range[j]) {
+				encode[j] = NOT_INCLUDE;
+			} else {
+				encode[j] = encode[j] + 1;
+				return true;
+			}
+		}
+		return false;
+	}
+
+	private void add(int[] encode, List<List<E>> inputs) {
+		List<E> tuple = decodeTuple(encode, inputs);
+		if (tuple.size() > argN || constraintCheck(tuple) == EvaluationResult.FALSE)
+			return;
+		fTupleOccurences.put(tuple, (tuple.size() == argN ? 1 : 0));
+	}
+
+	private void calculateFrequency() {
+		for (int m = argN - 1; m > 0; m--) {
+			List<List<E>> remove = new ArrayList<>();
+			for (List<E> key : fTupleOccurences.keySet()) {
+				if (key.size() == m) {
+					int freq = (int) fTupleOccurences.keySet().stream()
+							.filter(s -> s.size() == argN && s.containsAll(key)).count();
+					if (freq > 0)
+						fTupleOccurences.put(key, freq);
+					else
+						remove.add(key);
+				}
+			}
+			remove.forEach(s -> fTupleOccurences.remove(s));
+		}
+	}
+
+	private List<E> decodeTuple(int[] encode, List<List<E>> inputs) {
+		return IntStream.range(0, encode.length).mapToObj(i -> encode[i] == -1 ? null : inputs.get(i).get(encode[i]))
+				.filter(Objects::nonNull).collect(Collectors.toList());
+	}
+
+	private void calculateScore(int size) {
+		IntStream.range(1, size + 1).forEach(l -> {
+			fTupleOccurences.keySet().forEach(k -> {
+				if (k.size() == l) {
+					int update = fTupleOccurences.get(k);
+					if (l > 1)
+						update = calculateScore(k);
+					fScores.put(k, update);
+				}
+			});
+		});
+	}
+
+	private int calculateScore(List<E> tuple) {
+		return fTupleOccurences.get(tuple) * fTupleOccurences.keySet().stream()
+				.filter(k -> tuple.containsAll(k) && k.size() == tuple.size() - 1).mapToInt(fScores::get).sum();
+	}
+
+	public EvaluationResult constraintCheck(List<E> tuple) {
+		return fconstraintEvaluator.evaluate(tuple);
+	}
+
+	public int getNumOfTuples() {
+		return fTupleOccurences.size();
+	}
+
+	public void printSequence() {
+		for (int m = argN; m > 0; m--) {
+			int finalM = m;
+			fTupleOccurences.keySet().stream().filter(s -> s.size() == finalM).forEach(s -> {
+				s.forEach(k -> System.out.print(k + ","));
+				System.out.println(fTupleOccurences.get(s));
+			});
+		}
+	}
+
+	public int getScore(List<E> tuple) {
+		if (fScores.containsKey(tuple))
+			return fScores.get(tuple);
+		return -1;
+	}
+	
+	public void update(List<E> test) {
+		// if the test does not cover any argN-tuple, the scores does not update
+		if (test.size() < argN)
+			return;
+		// obtain all the argN-tuples covered by the test and remove them
+		fTupleOccurences.entrySet().removeIf(e -> e.getKey().size() == argN && test.containsAll(e.getKey()));
+		// update the occurences (i.e., fTupleOccurences) in the constructed table and scores (i.e., fScores)
+		calculateFrequency();
```

Comment:
This definitely can be optimized. You do not have to recalculate all the frequencies - you simply decrement values for all subtuples from the tuple above